core/src/main/java/com/exadel/aem/toolkit/core/injectors/utils/CastUtil.java

```diff
@@ -67,27 +67,31 @@ public static Object toType(Object value, Type type) {
      *                  object or primitive, the {@code converter} processes it as is
      * @return Transformed value; returns null in case type casting is not possible
      */
-    public static Object toType(Object value, Type type, BiFunction<Object, Type, Object> converter) {
+    public static CastResult toType(Object value, Type type, BiFunction<Object, Type, Object> converter) {
         if (value == null) {
             return null;
         }
 
+        if (value.getClass().isArray() && Array.getLength(value) == 0) {
```

Comment:
I didn't get the meaning of this addition. 
The idea of `CastResult`, is that `CastUtil` returns exactly the same value it used to. But it also returns the flag saying if it is a "naturally obtained" value, or a fallback value. (If it were Python, it would have been not a `CastResult` class at all, but a tuple of `Object, boolean`.)
Fallback value must never be `null`. We fight against nulls. Fallback value is **not null**, and **therefore** we need to mark it as "fallback", so that BaseInjector had its chance to apply a `@Default`.

So I think, instead of these lines, you need to "CastResult"-ify the return values of the many private methods of the current class in which default values are really applied. For example, the `toNumeric` method. It is where, if the `value` argument is `null`, we return `0`. And it is exactly where we must additionally say: this `0` is fallback! Same applies to arrays and collections, I believe

Reply:
fixed